### PR TITLE
Fix geolocation even if anonymize_ip is enabled

### PR DESCRIPTION
**Pull request by kuzmany**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Noticed If Anonymize IP  is enabled not geolocation provide. 
This PR fixed it. 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Enable Anonymize IP in configuration > Tracking Settings
2. Setup tracking code and go to incognito window  and hit the trakcing page
3. You contact should be IP address in anonymous  contact list tab without country flag.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Repeat all steps
3. You contact should listed with country flag

### CODE REVIEW

**Comment by kuzmany** [comment]
Did you test it on mautibox? I bet you forgot download IP lookup data in configuration . Now should works

https://mautibox.com/8175/s/contacts

image

File: app/bundles/LeadBundle/Model/LeadModel.php [hunk 1]
```diff
@@ -507,6 +507,10 @@ public function saveEntity($entity, $unlock = true)
             $details = $ips->first()->getIpDetails();
             // Only update with IP details if none of the following are set to prevent wrong combinations
             if (empty($fields['core']['city']['value']) && empty($fields['core']['state']['value']) && empty($fields['core']['country']['value']) && empty($fields['core']['zipcode']['value'])) {
+                if ($this->coreParametersHelper->getParameter('anonymize_ip') && $this->ipLookupHelper->getRealIp()) {
+                    $details = $this->ipLookupHelper->getIpDetails($this->ipLookupHelper->getRealIp());
```
Comment (Enc3phale) [inline app/bundles/LeadBundle/Model/LeadModel.php]: On this call if `$this->ipLookup == NULL` within `getIpDetails` function an error is thrown.
`Call to a member function setIpAddress() on null`

**Comment by kuzmany** [comment]
@Enc3phale last commit should fixed

File: app/bundles/LeadBundle/Model/LeadModel.php (hunk 1 above)
Comment (Enc3phale) [reply app/bundles/LeadBundle/Model/LeadModel.php]: With last commit https://github.com/mautic/mautic/pull/8175/commits/aa3139c8c9792aed335fbc1e176a8f007635a2aa, `$details from line 507` is erased by an empty array `if $this->ipLookup == NULL`, is this wanted behavior?

Comment (kuzmany) [reply app/bundles/LeadBundle/Model/LeadModel.php]: If ipLookup is null then there is fatal error. Do you have any experience?
But I noticed another warning, I will check it today

https://github.com/mautic/mautic/pull/8175/files#diff-680a217018ae349738fa38e34b7a19f8R228

File: app/bundles/LeadBundle/Model/LeadModel.php [hunk 2]
```diff
@@ -507,6 +507,10 @@ public function saveEntity($entity, $unlock = true)
             $details = $ips->first()->getIpDetails();
             // Only update with IP details if none of the following are set to prevent wrong combinations
             if (empty($fields['core']['city']['value']) && empty($fields['core']['state']['value']) && empty($fields['core']['country']['value']) && empty($fields['core']['zipcode']['value'])) {
+                if ($this->coreParametersHelper->getParameter('anonymize_ip') && $this->ipLookupHelper->getRealIp()) {
```
Comment (Enc3phale) [inline app/bundles/LeadBundle/Model/LeadModel.php]: What about this ? No fatal and not erased if ipLookupDetails is empty.
```suggestion
                if ($this->coreParametersHelper->getParameter('anonymize_ip') && $this->ipLookupHelper->getRealIp()) {
                    $ipLookupDetails = $this->ipLookupHelper->getIpDetails($this->ipLookupHelper->getRealIp());
                    
                    if (!empty($ipLookupDetails)) {
                        $details = $ipLookupDetails;
                    }
                }    
```

File: app/bundles/LeadBundle/Model/LeadModel.php (hunk 1 above)
Comment (kuzmany) [reply app/bundles/LeadBundle/Model/LeadModel.php]: But there is not reason $details contain something. 
If we use anonymize_ip then we don't have store any details if IP address (127.0.0.*** etc).
This reason why we ask details of real IP in this step

Comment (Enc3phale) [reply app/bundles/LeadBundle/Model/LeadModel.php]: Ok, perfect.

**Comment by npracht** [comment]
@dennisameling in case you have time, that could be added to 2.16.0 too i think ;)

**Comment by dennisameling** [comment]
@npracht Nice one, but the tests are failing ;-) Since I need to go ahead and release the 2.16 beta today, I'm skipping this one for 2.16.

**Comment by RCheesley** [comment]
@kuzmany can you review the failing Travis issue so we can maybe aim to get this into 2.16.1?

**Comment by dennisameling** [comment]
Closing and re-opening this PR to see if tests are still failing (we've had quite some commits after this PR was created). Just want to be sure here.

**Comment by dennisameling** [comment]
Unfortunately, tests are still failing. @kuzmany would you mind having a look? Thanks 😊

**Pull request merged**